### PR TITLE
RFC First cut support for sandboxing procmacros with wasm

### PR DESCRIPTION
**Pull request by jsgf**
This is a first cut experiment at building proc-macros with wasm in order to sandbox them.

It adds a new `Kind`: `HostSandbox`, in addition to `Host` and `Target`. This is treated as a host build, but its always `wasm32-unknown-unknown`. 

In Cargo.toml, you can set `wasm_sandbox = true` in addition to `proc-macro = true`, which enables the use of `HostSandbox`.

When the compilation actually happens, it's treated as a cross-build to wasm32-unknown-unknown.

Lightly tested on `serde`, which worked until it couldn't find `proc_macro` from rustc (well, proc-macro2's build.rs assumes it can't exist and so fakes it out).

Assuming this basic approach is sound, I'm thinking that the `wasm_sandbox` flag will become a marker that the procmacro *could* be sandboxed, and then a command-line or cargo config option would request that procmacros be built sandboxed (with further config to forbid non-sandboxed).

TODO:
- the rustc side
- what happens if we're cross-compiling to wasm32? Does it conflict or is it ok?

cc @alexcrichton, @dtolnay

### CODE REVIEW

**Comment by jsgf** [comment]
If you assume features are additive then you could just set them all for prebuilt. Or if she are rare and bloaty then prebuilt could exclude them and you'd need to build locally if you want them. 

IOW the prebuilt would have associated features and resolution would work out whether it's usable.

Of course it breaks if features are used wrongly.

File: src/cargo/util/toml/mod.rs [hunk 1]
```diff
@@ -1517,6 +1517,7 @@ struct TomlTarget {
     proc_macro: Option<bool>,
     #[serde(rename = "proc_macro")]
     proc_macro2: Option<bool>,
+    wasm_sandbox: Option<bool>,
```
Comment (killercup) [inline src/cargo/util/toml/mod.rs]:
```suggestion
    #[serde(rename = "wasm-sandbox")]
    wasm_sandbox: Option<bool>,
```

for consistency 😉

**Comment by jsgf** [comment]
@joshtriplett Totally on board with that characterization.

**Comment by alexcrichton** [comment]
Ok this has been quiet for some time now and in the meantime we've had [`watt`](https://github.com/rust-lang/rfcs/pull/2474) emerge as well as a way of testing this. Naturally this is going to be a pretty major feature and I think we'll probably want to start with an RFC before the implementation here as well.

In any case we're certainly interested in continuing to investigate this as always, and we're more than willing to help discuss this in the future!

**Comment by auscompgeek** [comment]
@alexcrichton that links to rfc 2474, which doesn't seem right given the context. Did you mean to link to something else?

**Comment by dtolnay** [comment]
Here is the link: https://github.com/dtolnay/watt

**Comment by alexcrichton** [comment]
Er yes sorry I apologize, @dtolnay has the right link.

**Pull request closed without merging**